gusto/advection.py

```diff
@@ -1,34 +1,97 @@
+    DG advection.
+    """
+    def get_apply(self, x_in, x_out):
+        if hasattr(self, "Projector"):
+            def new_apply(self, x_in, x_out):
+                self.xdg_in.interpolate(x_in)
+                original_apply(self, self.xdg_in, self.xdg_out)
+                self.Projector.project()
+                x_out.assign(self.x_projected)
+            return new_apply(self, x_in, x_out)
+
+        else:
+            return original_apply(self, x_in, x_out)
+    return get_apply
 
 
 class Advection(object):
     """
-    Base class for advection schemes for Gusto.
+    Base class for advection schemes.
 
     :arg state: :class:`.State` object.
+    :arg field: field to be advected
+    :arg equation: :class:`.Equation` object, specifying the equation
+    that field satisfies
+    :arg solver_params: solver_parameters
     """
     __metaclass__ = ABCMeta
 
-    def __init__(self, state):
+    def __init__(self, state, field, equation=None, solver_params=None):
+
         self.state = state
-        self.ubar = Function(state.V[0])
+        self.field = field
+
+        if solver_params is None:
+            if isinstance(equation, LinearAdvection):
+                self.solver_parameters = {'ksp_type':'cg',
+                                          'pc_type':'bjacobi',
+                                          'sub_pc_type': 'ilu'}
+            else:
+                self.solver_parameters = {'ksp_type':'preonly',
+                                          'pc_type':'bjacobi',
+                                          'sub_pc_type': 'ilu'}
+        else:
+            self.solver_parameters = solver_params
+        self.dt = self.state.timestepping.dt
+
+        # check to see if we are using an embedded DG method - is we are then
+        # the projector and output function will have been set up in the
+        # equation class and we can get the correct function space from
+        # the output function.
+        if hasattr(equation, "Projector"):
+            fs = equation.xdg_out.function_space()
+            self.Projector = equation.Projector
+            self.xdg_out = equation.xdg_out
+            self.xdg_in = Function(fs)
+            self.x_projected = equation.x_projected
+        else:
+            fs = field.function_space()
+
+        # setup required functions
+        self.dq = Function(fs)
+        self.q1 = Function(fs)
+
+        # get ubar from the equation class if provided
+        self.equation = equation
+        if equation is not None:
+            self.ubar = self.equation.ubar
+
+    def update_solver(self):
+        # setup solver using lhs and rhs defined in derived class
+
+        problem = LinearVariationalProblem(self.lhs, self.rhs, self.dq)
+        solver_name = self.field.name()+self.equation.__class__.__name__+self.__class__.__name__
+        self.solver = LinearVariationalSolver(problem, solver_parameters=self.solver_parameters, options_prefix=solver_name)
```

Comment:
I would be tempted to make this as a `cached_property`:

```python
from firedrake.utils import cached_property

class Foo(...):
    
    @cached_property
    def solver(self):
        problem = ...
        name = ...
        return LinearVariationalSolver(problem, ...)
```

At the first reference of `self.solver`, the solver is created.  It is then subsequently cached on the object, so subsequent references just return the same thing.  Then the subclasses don't need to call `update_solver` ever.  It is just made when needed.